extensions/extensionservice/org.eclipse.smarthome.extensionservice.marketplace/src/main/java/org/eclipse/smarthome/extensionservice/marketplace/internal/MarketplaceExtensionService.java

```diff
@@ -0,0 +1,303 @@
+                if (handler.supports(mpExt)) {
+                    if (!handler.isInstalled(mpExt)) {
+                        try {
+                            handler.install(mpExt);
+                            postInstalledEvent(extensionId);
+                        } catch (MarketplaceHandlerException e) {
+                            postFailureEvent(extensionId, e.getMessage());
+                        }
+                    } else {
+                        postFailureEvent(extensionId, "Extension is already installed.");
+                    }
+                    return;
+                }
+            }
+        }
+        postFailureEvent(extensionId, "Extension not known.");
+    }
+
+    @Override
+    public void uninstall(String extensionId) {
+        Extension ext = getExtension(extensionId, null);
+        if (ext instanceof MarketplaceExtension) {
+            MarketplaceExtension mpExt = (MarketplaceExtension) ext;
+            for (MarketplaceExtensionHandler handler : extensionHandlers) {
+                if (handler.supports(mpExt)) {
+                    if (handler.isInstalled(mpExt)) {
+                        try {
+                            handler.uninstall(mpExt);
+                            postUninstalledEvent(extensionId);
+                        } catch (MarketplaceHandlerException e) {
+                            postFailureEvent(extensionId, e.getMessage());
+                        }
+                    } else {
+                        postFailureEvent(extensionId, "Extension is not installed.");
+                    }
+                    return;
+                }
+            }
+        }
+        postFailureEvent(extensionId, "Extension not known.");
+    }
+
+    private void postInstalledEvent(String extensionId) {
+        Event event = ExtensionEventFactory.createExtensionInstalledEvent(extensionId);
+        eventPublisher.post(event);
+    }
+
+    private void postUninstalledEvent(String extensionId) {
+        Event event = ExtensionEventFactory.createExtensionUninstalledEvent(extensionId);
+        eventPublisher.post(event);
+    }
+
+    private void postFailureEvent(String extensionId, String msg) {
+        Event event = ExtensionEventFactory.createExtensionFailureEvent(extensionId, msg);
+        eventPublisher.post(event);
+
+    }
+
+    private String getExtensionId(Node node) {
+        StringBuilder sb = new StringBuilder(MarketplaceExtension.EXT_PREFIX);
+        switch (node.packagetypes) {
+            case MP_PACKAGETYPE_RULE_TEMPLATE:
+                sb.append(MarketplaceExtension.EXT_TYPE_RULE_TEMPLATE).append("-");
+                break;
+            case MP_PACKAGETYPE_BINDING:
+                sb.append(MarketplaceExtension.EXT_TYPE_BINDING).append("-");
+                break;
+            default:
+                return null;
+        }
+        sb.append(node.id.replaceAll("[^a-zA-Z0-9_]", ""));
```

Comment:
We are adding the "marketplace:" prefix for the extension ids and hence the part AFTER that prefix should not contain any additional colons. What is converted here is the node id and this is (from what I have seen so far) always a numerical number anyhow, so this line is only a safeguard for the future.